### PR TITLE
Migrate the ssh tunnel from supervisor to ws-proxy

### PR DESCRIPTION
**Pull request by iQQBot**
## Description

Migrate the SSH tunnel from supervisor to ws-proxy, detail in ENG-1237

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at dc0da1a</samp>

This pull request adds SSH over websocket tunnel support to the ws-proxy component, which allows SSH access to workspaces over a websocket connection. It also refactors the proxy and sshproxy packages to use a common package for shared types and constants related to workspace information.

</details>

## Related Issue(s)

Fixes ENG-1237

## How to test

1. test old ssh gateway behavior not change
2. test with vscode desktop extension, make sure it doesn't break
3. test with latest jetbrains gateway plugin [link](https://plugins.jetbrains.com/plugin/18438-gitpod-gateway/versions/dev), and set `Force SSH over HTTP tunnel` to `true` and test Jetbrains IDE, make sure it doesn't break 
4. write a scrip to test with wrong ownerToken, make sure it cannot connect to tunnel

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - pd-ssh-ove75da47021d</li>
	<li><b>🔗 URL</b> - <a href="https://pd-ssh-ove75da47021d.preview.gitpod-dev.com/workspaces" target="_blank">pd-ssh-ove75da47021d.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - pd-ssh-over-websocket-gha.20630</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-pd-ssh-ove75da47021d%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [x] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/ws-proxy/pkg/sshproxy/server.go [hunk 1]
```diff
@@ -106,19 +108,45 @@ func init() {
 
 // New creates a new SSH proxy server
 
-func New(signers []ssh.Signer, workspaceInfoProvider proxy.WorkspaceInfoProvider, heartbeat Heartbeat) *Server {
+func New(signers []ssh.Signer, workspaceInfoProvider common.WorkspaceInfoProvider, heartbeat Heartbeat) *Server {
```
Comment (akosyakov) [inline components/ws-proxy/pkg/sshproxy/server.go:111]: Does it mean we now also will get metrics for SSH tunnel connections from ws-proxy?

**Comment by akosyakov** [comment]
Tagged @gitpod-io/team-experience since we need to validate that our SSH integrations are still working.

**Comment by mustard-mh** [comment]
|JetBrains Editor (websocket tunnel)| VSCode Desktop (local-ssh) |
|:-:|:-:|
|image|image|

**Comment by mustard-mh** [comment]
Wrong ownerToken

image

**Comment by mustard-mh** [comment]
@iQQBot Did I miss any case?

**Comment by mustard-mh** [comment]
Tested with localssh, wrong ownerToken

image

**Review by mustard-mh — APPROVED** [review]

**Review by aledbf — APPROVED** [review]

**Pull request merged**